celavi/pylca_celavi/des_interface.py

```diff
@@ -193,67 +193,74 @@ def pylca_run_main(self, df):
-                # Calculates the concrete lifecycle flow and emissions inventory
-                df_static,df_emissions = concrete_life_cycle_inventory_updater(new_df, year, material, stage, self.static_lci_filename, self.stock_filename, self.emissions_lci_filename, self.substitution_rate)
-    
-                if not df_static.empty:
-    
-                    working_df = df_with_no_lca_entry
-                    working_df['flow name'] = working_df['material'] + ', ' + working_df['stage']
-                    working_df= working_df[['flow name','flow quantity']]
-    
-                    # model_celavi_lci() is calculating foreground processes and dynamics of electricity mix.
-                    # It calculates the LCI flows of the foreground process.
-                    res = model_celavi_lci(working_df,year,facility_id,stage,material,df_static,self.dynamic_lci_filename,self.intermediate_demand_filename)
+            
+        
+            #This function breaks down the df sent from DES to individual rows with unique rows, facilityID, stage and materials.
+            for index,row in df_s.iterrows():
+                
+                year = row['year']
+                stage = row['stage']
+                material = row['material']
+                facility_id = row['facility_id']
+                state = row['state']
+                new_df = df_s[df_s['index'] == index]
     
-                    # model_celavi_lci_insitu() calculating direct emissions from foreground
-                    # processes.
-                    emission = model_celavi_lci_insitu(working_df,year,facility_id,stage,material,df_emissions)
+                if self.use_shortcut_lca_calculations:
+                    #Calling the lca performance improvement function to do shortcut calculations. 
+                    df_with_no_lca_entry,result_shortcut = self.lca_performance_improvement(new_df)
+        
+                else:
+                    
+                    df_with_no_lca_entry = new_df
+                    result_shortcut = pd.DataFrame()
     
-                    if not res.empty:
-                        res = model_celavi_lci_background(res,year,facility_id,stage,material,self.uslci_filename,self.lci_activity_locations)
-                        lci = postprocessing(res,emission)
-                        res = impact_calculations(lci,self.traci_lci_filename)
-                        res_df = pd.concat([res_df,res])
-                        lcia_mass_flow = pd.concat([lci,lcia_mass_flow])
-                        
+                if not df_with_no_lca_entry.empty:
+                    # Calculates the concrete lifecycle flow and emissions inventory
+                    df_static,df_emissions = concrete_life_cycle_inventory_updater(new_df, year, material, stage, self.static_lci_filename, self.stock_filename, self.emissions_lci_filename, self.substitution_rate)
+        
+                    if not df_static.empty:
```

Comment:
Whenever an empty results data frame signifies an error with the optimization, add in a warning that we can use for debugging. There's a few statements like this throughout the code.

Reply:
Added statements for optimization failure warning.